src/psyclone/f2pygen.py

```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
-        '''
+    '''
+    # The Fortran intrinsic types supported by this class
+    SUPPORTED_TYPES = ["integer", "real", "logical"]
+
+    def __init__(self, parent, datatype="", entity_decls=None, intent="",
+                 pointer=False, kind="", dimension="", allocatable=False,
+                 save=False, initial_values=None):
         if entity_decls is None:
             raise RuntimeError(
                 "Cannot create a variable declaration without specifying the "
                 "name(s) of the variable(s)")
         fort_fmt = FortranFormat(True, False)  # free form, strict
-        if datatype.lower() == "integer":
+
+        dtype = datatype.lower()
+        if dtype not in self.SUPPORTED_TYPES:
+            raise RuntimeError(
+                "f2pygen.DeclGen.init: Only {0} types are currently"
+                " supported and you specified '{1}'"\
+                .format(self.SUPPORTED_TYPES, datatype))
+
+        # If initial values have been supplied then check that there
+        # are the right number of them and that they are consistent
+        # with the type of the variable(s) being declared.
+        if initial_values:
+            if len(initial_values) != len(entity_decls):
+                raise RuntimeError(
+                    "f2pygen.DeclGen.init: number of initial values supplied "
+                    "({0}) does not match the number of variables to be "
+                    "declared ({1}: {2})".format(len(initial_values),
+                                                 len(entity_decls),
+                                                 str(entity_decls)))
+            self._check_initial_values(dtype, initial_values)
+
+        if dtype == "integer":
             reader = FortranStringReader("integer :: vanilla")
             reader.set_format(fort_fmt)
             myline = reader.next()
             self._decl = fparser1.typedecl_statements.Integer(parent.root,
                                                               myline)
-        elif datatype.lower() == "real":
+        elif dtype == "real":
             reader = FortranStringReader("real :: vanilla")
             reader.set_format(fort_fmt)
             myline = reader.next()
             self._decl = fparser1.typedecl_statements.Real(parent.root, myline)
+        elif dtype == "logical":
+            reader = FortranStringReader("logical :: vanilla")
+            reader.set_format(fort_fmt)
+            myline = reader.next()
+            self._decl = fparser1.typedecl_statements.Logical(parent.root,
+                                                              myline)
         else:
-            raise RuntimeError(
-                "f2pygen:DeclGen:init: Only integer and real are currently"
-                " supported and you specified '{0}'".format(datatype))
-        # make a copy of entity_decls as we may modify it
+            # Defensive programming in case SUPPORTED_TYPES is added to
```

Comment:
Good. We should always include such checks in if statements as it makes adding new functionality so much easier.